### PR TITLE
[JSC] Inline extra-stack-slot filling path in arity-fixup

### PR DESCRIPTION
**Pull request by Constellation**
#### 747ff0a3538d54a2550a5f43bc31c6459e9e8040
<pre>
[JSC] Inline extra-stack-slot filling path in arity-fixup
<a href="https://bugs.webkit.org/show_bug.cgi?id=278050">https://bugs.webkit.org/show_bug.cgi?id=278050</a>
<a href="https://rdar.apple.com/problem/133791041">rdar://problem/133791041</a>

Reviewed by NOBODY (OOPS!).

Stack is always aligned with 16-bytes. As a result, JS arguments may have one extra stack slot sometimes. Like, for example,
if passed argument is only one, then one slot remains unused. And when arity fixup happens, and the expected arguments are two,
then we just fill this additional one slot with undefined and finishes fixup. In this case we do not need to modify sp at all.
In this patch, we inline this path in the caller side of arity fixup function. The code is not modifying sp at all.

* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
(JSC::DFG::SpeculativeJIT::compileFunction):
* Source/JavaScriptCore/ftl/FTLCompile.cpp:
(JSC::FTL::compile):
* Source/JavaScriptCore/jit/JIT.cpp:
(JSC::JIT::compileAndLinkWithoutFinalizing):
* Source/JavaScriptCore/llint/LLIntThunks.cpp:
(JSC::LLInt::arityFixupThunk):
</pre>
https://github.com/WebKit/WebKit/commit/747ff0a3538d54a2550a5f43bc31c6459e9e8040

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62537 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/41892 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15132 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/66521 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/13089 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/49579 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13425 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/50391 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9019 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/65606 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38922 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54162 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31131 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/35639 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/11490 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12017 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/55634 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/57243 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/11806 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/68250 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/61780 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/6483 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/11499 "Found 1 new test failure: workers/worker-user-gesture.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/57764 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/6512 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54202 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57954 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5394 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/83543 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/37692 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14673 "Found 1 new JSC stress test failure: wasm.yaml/wasm/fuzz/memory.js.wasm-eager (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/38777 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/39874 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/38521 "Built successfully") | | |

### CODE REVIEW

File: Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp [hunk 1]
```diff
@@ -236,6 +236,16 @@ void SpeculativeJIT::compileFunction()
         branch32(AboveOrEqual, GPRInfo::argumentGPR2, TrustedImm32(numberOfParameters)).linkTo(entryLabel, this);
 
         getArityPadding(vm(), numberOfParameters, GPRInfo::argumentGPR2, GPRInfo::argumentGPR0, GPRInfo::argumentGPR1, GPRInfo::argumentGPR3, stackOverflowWithEntry);
+        subPtr(stackPointerRegister, TrustedImm32(static_cast<int32_t>(sizeof(CallerFrameAndPC))), GPRInfo::regT3); // Initially expected callFramePointer after prologue.
+        add32(TrustedImm32(CallFrame::headerSizeInRegisters), GPRInfo::argumentGPR2);
+
+        static_assert(stackAlignmentRegisters() == 2);
```
Comment (hyjorc1) [inline Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:246]: Maybe add a comment `Check to see if we have extra slots we can use` just like the previous one.

File: Source/JavaScriptCore/ftl/FTLCompile.cpp [hunk 2]
```diff
@@ -157,6 +157,16 @@ void compile(State& state, Safepoint::Result& safepointResult)
             unsigned numberOfParameters = codeBlock->numParameters();
             CCallHelpers::JumpList stackOverflowWithEntry;
             jit.getArityPadding(vm, numberOfParameters, GPRInfo::argumentGPR2, GPRInfo::argumentGPR0, GPRInfo::argumentGPR1, GPRInfo::argumentGPR3, stackOverflowWithEntry);
+            jit.subPtr(CCallHelpers::stackPointerRegister, CCallHelpers::TrustedImm32(static_cast<int32_t>(sizeof(CallerFrameAndPC))), GPRInfo::regT3); // Initially expected callFramePointer after prologue.
+            jit.add32(CCallHelpers::TrustedImm32(CallFrame::headerSizeInRegisters), GPRInfo::argumentGPR2);
+
+            static_assert(stackAlignmentRegisters() == 2);
```
Comment (hyjorc1) [inline Source/JavaScriptCore/ftl/FTLCompile.cpp:167]: Ditto.

**Review by hyjorc1 — APPROVED** [review]
r=me

File: Source/JavaScriptCore/ftl/FTLCompile.cpp (hunk 2 above)
Comment (Constellation) [reply Source/JavaScriptCore/ftl/FTLCompile.cpp:167]: Sounds good

File: Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp (hunk 1 above)
Comment (Constellation) [reply Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:246]: Sounds good

**Comment by Constellation** [comment]
It was not much beneficial.

**Pull request closed without merging**